components/homme/src/share/cxx/SphereOperators.hpp

```diff
@@ -1176,6 +1176,273 @@ class SphereOperators
   Real m_scale_factor_inv, m_laplacian_rigid_factor;
 };
 
+using TeamPolicy = Kokkos::TeamPolicy<ExecSpace>;
+using Team = TeamPolicy::member_type;
+
+static constexpr int NPNP = NP * NP;
+static constexpr int WARP_SIZE = warpSize;
+
+static constexpr int SPHERE_BLOCK_LEV = WARP_SIZE;
+static constexpr int SPHERE_BLOCKS_PER_COL = (NUM_PHYSICAL_LEV - 1) / SPHERE_BLOCK_LEV + 1;
+static constexpr int SPHERE_BLOCK = NPNP * SPHERE_BLOCK_LEV;
+
+struct SphereGlobal {
+
+  const ExecViewManaged<const Real *[2][2][NP][NP]> d;
+  const ExecViewManaged<const Real *[2][2][NP][NP]> dinv;
+  const ExecViewManaged<const Real [NP][NP]> dvv;
+  const ExecViewManaged<const Real *[NP][NP]> metdet;
+  const Real scale_factor_inv;
+
+  SphereGlobal(const SphereOperators &op):
+    d(op.m_d),
+    dinv(op.m_dinv),
+    dvv(op.dvv),
+    metdet(op.m_metdet),
+    scale_factor_inv(op.m_scale_factor_inv)
+  {}
+};
+
+using SphereBlockScratchView = Kokkos::View<
+  Real[SPHERE_BLOCK_LEV][NP][NP],
+  ExecSpace::scratch_memory_space,
+  Kokkos::MemoryTraits<Kokkos::Unmanaged>
+  >;
+
+using SphereBlockScratchSubview = Kokkos::Subview<
+  SphereBlockScratchView, int,
+  std::remove_const_t<decltype(Kokkos::ALL)>,
+  std::remove_const_t<decltype(Kokkos::ALL)>
+  >;
+
+struct SphereBlockOps;
+
+struct SphereBlockScratch {
+  SphereBlockScratchView v;
+  SphereBlockScratchSubview sv;
+  KOKKOS_INLINE_FUNCTION SphereBlockScratch(const SphereBlockOps &b);
+};
+
+struct SphereBlockOps {
+  const SphereGlobal &g;
+  const Team &t;
+  Real scale_factor_inv;
+  Real metdet;
+  Real rrdmd;
+  Real dinv[2][2];
+  Real dvvx[NP];
+  Real dvvy[NP];
+  int e,x,y,z;
+
+  KOKKOS_INLINE_FUNCTION SphereBlockOps(const SphereGlobal &sg, const Team &team):
+    g(sg),
+    t(team),
+    scale_factor_inv(g.scale_factor_inv),
+    rrdmd(0)
+  {
+    const int lr = t.league_rank();
+    e = lr / SPHERE_BLOCKS_PER_COL;
+    const int iw = lr % SPHERE_BLOCKS_PER_COL;
+    const int tr = t.team_rank();
+    const int ixy = tr / SPHERE_BLOCK_LEV;
+    x = ixy / NP;
+    y = ixy % NP;
+    const int dz = tr % SPHERE_BLOCK_LEV;
+    z = dz + iw * SPHERE_BLOCK_LEV;
+
+    metdet = g.metdet(e,x,y);
+    for (int i = 0; i < 2; i++) for (int j = 0; j < 2; j++) dinv[i][j] = g.dinv(e,i,j,x,y);
+    for (int j = 0; j < NP; j++) {
+      dvvx[j] = g.dvv(x,j);
+      dvvy[j] = g.dvv(y,j);
+    }
+  }
+
+  KOKKOS_INLINE_FUNCTION void barrier() const
+  {
+    t.team_barrier();
+  }
+
+  KOKKOS_INLINE_FUNCTION Real div(const SphereBlockScratch &t0, const SphereBlockScratch &t1)
+  {
+    // Separately compute du and dv to match Fortran bfb
+    Real du = 0;
+    Real dv = 0;
+    for (int j = 0; j < NP; j++) {
+      du += dvvy[j] * t0.sv(x,j);
+      dv += dvvx[j] * t1.sv(j,y);
+    }
+    if (rrdmd == 0) rrdmd = (1.0 / metdet) * scale_factor_inv;
```

Comment:
Divisions are expensive on AMD GPUs. The attempt here is to do it only once per kernel, even if the kernel calls `div` or `fort` more than once, or both `div` and `vort`, and to do it zero times if the kernel never calls `div` or `vort` (but calls only `grad`, for example).

Reply:
Want me to try converting to a stored `rmetdet` and see how performance compares?

Reply:
No need for that. To avoid doing the division at every kernel call, we'd have to precompute the rmetdet view during init, and storing in the ElementsGeometry structure. I don't think that's a change for this PR. I was just thinking out loud :)